wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/kubernetes/KubernetesToolToWorkspaceApplier.java

```diff
@@ -192,26 +204,27 @@ private void estimateCommandsMachineName(
     toolCommands.forEach(c -> c.getAttributes().put(MACHINE_NAME_ATTRIBUTE, machineName));
   }
 
-  private KubernetesList unmarshal(Tool tool, String recipeContent)
+  private List<HasMetadata> unmarshal(Tool tool, String recipeContent)
       throws DevfileRecipeFormatException {
     try {
-      return Serialization.unmarshal(recipeContent, KubernetesList.class);
-    } catch (KubernetesClientException e) {
+      return objectsParser.parse(recipeContent);
+    } catch (Exception e) {
       throw new DevfileRecipeFormatException(
           format(
               "Error occurred during parsing list from file %s for tool '%s': %s",
               tool.getLocal(), tool.getName(), e.getMessage()));
     }
   }
 
-  private String asYaml(Tool tool, KubernetesList list) throws DevfileRecipeFormatException {
+  private String asYaml(Tool tool, List<HasMetadata> list) throws DevfileRecipeFormatException {
     try {
-      return Serialization.asYaml(list);
+      return Serialization.asYaml(new KubernetesListBuilder().withItems(list).build());
```

Comment:
Am I right in thinking that this will parse a Template devfile and then serialize it to a List? If so, can this be avoided?

Reply:
You're right. Any of single object(like deployment), list, a template will be serialized as a list with processed objects.
In the simplest case - when Devfile contains one kubernetes or openshift tool we could avoid it, and serialize the same object that tool contains.
But if Devfile contains two kubernetes/openshift tools then they may have different objects (one tool has a single object, another has a template and another tool has a list of objects). In such case, we can not serialize the same object as we got (I tried to serialize different tools objects (list, template, single object) as a separate item of wrapper list BUT fabric8 client does not support list as an item of a list).
So, I implemented it in this way because it makes things simpler and we're going to deprecate workspaceConfigs ( it is confirmed by Mario https://github.com/eclipse/che/issues/12719#issuecomment-467797983), no need to invest more time on it.
Hope, my explanation is detailed/clear enough, feel free to ask more if needed.